jax/api.py

```diff
@@ -983,12 +983,17 @@ def __repr__(self):
     return '<jax.custom_transforms function {fun}>'.format(fun=self.__name__)
 
   def __call__(self, *args, **kwargs):
+    def pv_like(x):
+      return pe.PartialVal((batching.get_aval(x), core.unit))  # Use shaped aval
     jax_args, in_trees = unzip2(map(pytree_to_jaxtupletree, args))
     jax_kwargs, kwargs_tree = pytree_to_jaxtupletree(kwargs)
-    out_tree = lu.Store()
-    ans = self.prim.bind(jax_kwargs, *jax_args, kwargs_tree=kwargs_tree,
-                         in_trees=in_trees, out_tree=out_tree)
-    return build_tree(out_tree.val, ans)
+    jaxtree_fun, out_tree = pytree_fun_to_jaxtupletree_fun2(
+        lu.wrap_init(self.fun), kwargs_tree, in_trees)
+    pvals_in = map(pv_like, (jax_kwargs,) + jax_args)
+    jaxpr, _, consts = pe.trace_to_jaxpr(jaxtree_fun, pvals_in, instantiate=True)
```

Comment:
Wasn't sure what the relevance of instantiate is here, and whether I need instantiate=True.

Reply:
`instantiate` basically means whether constants are lifted into the jaxpr. By default `make_jaxpr` currently uses the default value of `instantiate=False` so you can see behavior like this:

```python
In [3]: make_jaxpr(lambda x: np.array([1., 2.]))(3)
Out[3]:
{ lambda  ;  ; a.
  let
  in * }
```

Where'd the constant go? Since it was known at tracing time rather than having a data dependence on the argument, it's in the `out_pval`, which is returned from `trace_to_jaxpr` (but on the line you highlighted is being set to the underscore value). This is why we call [`merge_pvals` in xla.py](https://github.com/google/jax/blob/29d9400aba53427ddc811821649f86aec6406c94/jax/interpreters/xla.py#L633): that combines the known-at-trace-time values in the pval with the computed-in-xla values.

If we edit `make_jaxpr` to call `trace_to_jaxpr` with `instantiate=True` we instead see this:

```python
In [2]: make_jaxpr(lambda x: np.array([1., 2.]))(3)
Out[2]:
{ lambda b ;  ; a.
  let
  in b }
```

The known constant has been lifted into a jaxpr constant, so it's not in the pval anymore and you don't need to use `merge_pvals` (though in this case it has to be passed in as a const argument when evaluating the jaxpr).

Yet one more behavior is that jaxprs can have literals, so if we instead make the output a constant scalar rather than an array, we see this regardless of whether `instantiate=True` or `instantiate=False`:

```python
In [3]: make_jaxpr(lambda x: 2)(3)
Out[3]:
{ lambda  ;  ; a.
  let
  in 2 }
```

We only added literals recently, and we think we should probably get rid of constants now that we have them (because they're somewhat redundant). But we haven't done that yet.

Reply:
As to whether we want `instantiate=True` here, I think it's a good idea because then we don't have to worry about pvals.